File: basic/c#/ResilientDatasetProcessing.cs
using System;
using System.IO;

var lines = File.ReadAllLines("../../data/customers.csv");

int processedRows = 0;
int failedRows = 0;

foreach (var line in lines[1..])
{
    try
    {
        var columns = line.Split(',');

        var id = columns[0];
        var age = int.Parse(columns[1]);
        var income = double.Parse(columns[2]);

        var score = age * income;

        Console.WriteLine($"Processed customer {id} -> score={score}");
    }
    catch (Exception error)
    {
        failedRows++;

        Console.WriteLine($"Failed row -> {error.Message}");
    }
    finally
    {
        processedRows++;
    }
}

Console.WriteLine("\nSummary");
Console.WriteLine($"Processed rows: {processedRows}");
Console.WriteLine($"Failed rows: {failedRows}");

File: basic/python/basic/resilient_dataset_processing.py
import csv

processed_rows = 0
failed_rows = 0

with open("../../data/customers.csv", newline="") as file:
    reader = csv.DictReader(file)

    for row in reader:
        try:
            age = int(row["age"])
            income = float(row["monthly_income"])

            score = age * income

            print(f"Processed customer {row['id']} -> score={score}")

        except ValueError as error:
            failed_rows += 1
            print(f"Failed customer {row['id']} -> {error}")

        finally:
            processed_rows += 1

print("\nSummary")
print(f"Processed rows: {processed_rows}")
print(f"Failed rows: {failed_rows}")
